### ISSUE TITLE
Intermittent TypeInitializationException in the DiffEngine.ProcessCleanup class

### ISSUE THREAD

**Issue by 0xced** [VerifyTests/DiffEngine#265]
Using DiffEngine 8.5.1 (which, by the way, doesn't have a tag on this repository) through Verify.Xunit 14.9.1 I sometimes get this exception when running tests on macOS (in Rider, but that's not relevant).

```
System.TypeInitializationException
The type initializer for 'DiffEngine.ProcessCleanup' threw an exception.
   at DiffEngine.ProcessCleanup.Kill(String command)
   at DiffEngine.DiffRunner.Kill(String tempFile, String targetFile) in C:\projects\diffengine\src\DiffEngine\DiffRunner_Kill.cs:line 32
   at VerifyEngine.ProcessEquals() in /_/src/Verify/Verifier/VerifyEngine.cs:line 218
   at VerifyEngine.ThrowIfRequired(String message) in /_/src/Verify/Verifier/VerifyEngine.cs:line 117
   at InnerVerifier.VerifyInner(Object target, Func`1 cleanup, IEnumerable`1 targets) in /_/src/Verify/Verifier/InnerVerifier_Inner.cs:line 30
   at InnerVerifier.Verify[T](T target) in /_/src/Verify/Verifier/InnerVerifier_Json.cs:line 65
   at VerifyXunit.Verifier.<>c__DisplayClass1_0.<<Verify>b__0>d.MoveNext() in /_/src/Verify.Xunit/Verifier.cs:line 32
--- End of stack trace from previous location ---
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 264
--- End of stack trace from previous location ---
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in C:\Dev\xunit\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:line 90
```

The underlying exception is either one of those exception:

```
System.ArgumentOutOfRangeException
Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'chunkLength')
   at System.Text.StringBuilder.ToString()
   at System.Runtime.CompilerServices.DefaultInterpolatedStringHandler.AppendFormatted[T](T value)
   at LinuxOsxProcess.RunPs() in C:\projects\diffengine\src\DiffEngine\Process\LinuxOsxProcess.cs:line 108
   at LinuxOsxProcess.FindAll()+MoveNext() in C:\projects\diffengine\src\DiffEngine\Process\LinuxOsxProcess.cs:line 31
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at DiffEngine.ProcessCleanup.Refresh() in C:\projects\diffengine\src\DiffEngine\Process\ProcessCleanup.cs:line 37
   at DiffEngine.ProcessCleanup..cctor() in C:\projects\diffengine\src\DiffEngine\Process\ProcessCleanup.cs:line 30
```

or

```
System.Exception
Process timed out. Command line: ps -o pid,command -x.
Output:   PID COMMAND
  455 /System/Library/Frameworks/LocalAuthentication.framework/Support/coreauthd
  456 /usr/sbin/cfprefsd agent
  460 /usr/libexec/UserEventAgent (Aqua)
  462 /usr/sbin/distnoted agent
  465 /System/Library/PrivateFrameworks/CloudServices.framework/Helpers/com.apple.sbd
  466 /usr/libexec/knowledge-agent
  467 /usr/libexec/trustd --agent
  468 /usr/libexec/lsd
  469 /usr/libexec/secd
  470 /System/Library/PrivateFrameworks/CloudKitDaemon.framework/Support/cloudd
  […]
```

The `ArgumentOutOfRangeException` is baffling, is there a bug in the new [String Interpolation in C# 10 and .NET 6](https://devblogs.microsoft.com/dotnet/string-interpolation-in-c-10-and-net-6/)? 🤯 Note that this execption occurs [while creating the interpolated string](https://github.com/VerifyTests/DiffEngine/blob/c44272071bd895d6d99c1e77b3dd4d5f58fb0701/src/DiffEngine/Process/LinuxOsxProcess.cs#L108) serving as message of the `System.Exception`. Eventually it fails inside [StringBuilder.ToString()](https://github.com/dotnet/runtime/blob/v6.0.0/src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs#L369)!

Anyway, when the `ArgumentOutOfRangeException` does not occur, the `System.Exception` occurs because of `ps -o pid,command -x` timing out after 500 ms occurs.

500 ms is a really long time for the `ps` command to timeout but it occurs, I've witnessed it several times but it's unfortunately impossible to reproduce reliably. I have run a quick experiment with [hyperfine](https://github.com/sharkdp/hyperfine) and I was only able to measure `ps -o pid,command -x` to last 388.2 ms:

```
hyperfine -m 1000 'ps -o pid,command -x' 
Benchmark 1: ps -o pid,command -x
  Time (mean ± σ):      89.4 ms ±  23.1 ms    [User: 21.1 ms, System: 64.9 ms]
  Range (min … max):    64.8 ms … 388.2 ms    1000 runs
```

Maybe the best solution is simply to increase the timeout? But what would be a reasonable value? 1 second, more?

It's too bad the `System.Diagnostics.Process` did not get much love in .NET Core and that iterating over `Process.GetProcesses()` and accessing its `StartInfo` property throws `System.InvalidOperationException: Process was not started by this object, so requested information cannot be determined.`. Using `System.Diagnostics.Process` would have been a cleaner solution than having to rely on the `ps` command!

**Comment by SimonCropp** [comment]
have u raised a bug for the string builder issue?

**Comment by 0xced** [comment]
No, I haven't. I think that the non-reproducible stack trace that I have might not be enough to be actionable.

**Comment by SimonCropp** [comment]
can u try Verify 14.10.1

**Comment by 0xced** [comment]
Aborting instead of throwing will definitely help! I'm pretty confident I won't see this error again.

Thanks once again for the very quick response time.

**Comment by 0xced** [comment]
I investigated a bit and the `System.ArgumentOutOfRangeException` that I've described is **not** a bug in the .NET runtime. It might happen when the `ps` process has not exited and the string builders continue to be fed in the `OutputDataReceived` and `ErrorDataReceived` events.

Removing `outputBuilder` and `errorBuilder` from the trace message in 7bf7f14f08d3fc15627004eaf8a5e29eec294214 solves this issue. Another possibility would have been to call `process.CancelOutputRead()` and `process.CancelErrorRead()` when the `DoubleWaitForExit` returns false to ensure that the `outputBuilder` and `errorBuilder` won't be updated anymore.
